code/datums/components/payment.dm

```diff
@@ -17,29 +17,36 @@
 	var/transaction_style = "Clinical"
 	///Who's getting paid?
 	var/datum/bank_account/target_acc
+	///Does this payment component respect same-department-discount?
+	var/department_discount = FALSE
 
 /datum/component/payment/Initialize(_cost, _target, _style)
 	target_acc = _target
 	if(!target_acc)
 		target_acc = SSeconomy.get_dep_account(ACCOUNT_CIV)
-
 	cost = _cost
 	transaction_style = _style
 	RegisterSignal(parent, COMSIG_OBJ_ATTEMPT_CHARGE, .proc/attempt_charge)
 	RegisterSignal(parent, COMSIG_OBJ_ATTEMPT_CHARGE_CHANGE, .proc/change_cost)
 
 /datum/component/payment/proc/attempt_charge(datum/source, atom/movable/target, extra_fees = 0)
 	SIGNAL_HANDLER
-
-	if(!cost) //In case a free variant of anything is made it'll skip charging anyone.
+	if(!cost && !extra_fees) //In case a free variant of anything is made it'll skip charging anyone.
 		return
+	var/total_cost = cost + extra_fees
 	if(!ismob(target))
 		return COMPONENT_OBJ_CANCEL_CHARGE
 	var/mob/living/user = target
+	if(issilicon(user)) //They have evolved beyond the need for mere credits
+		return
 	var/obj/item/card/id/card
 	if(istype(user))
 		card = user.get_idcard(TRUE)
+	if(!card && istype(user.pulling, /obj/item/card/id))
+		card = user.pulling
 	if(!card)
+		if(handle_cardless(user, total_cost)) //Here we attempt to handle the purchase physically, with held money first. Otherwise we default to below.
+			return
 		switch(transaction_style)
 			if(PAYMENT_FRIENDLY)
 				to_chat(user, span_warning("ID not detected, sorry [user]!"))
```

Comment:
Also doesn't that mean these errors are a bit outdated?

Reply:
Current implementation means that it'll go through the tree for the user's held/pulled card first, then if it's successfully handled via cash/without a card, it has an early return and logging/behavior is effectively early-returned. Otherwise it goes down the tree here in attempt_charge.

I'll admit, this is written pretty messily. I could probably move the dialog trees over to datumized behavior styles to cut down on the clutter if you'd prefer.

Reply:
What happens if I try and pay with cash, but don't have enough money? won't I still get these?